ocaml/testsuite/tests/typing-labeled-tuples/labeled_tuple_patterns.ml

```diff
@@ -0,0 +1,521 @@
+|}]
+
+(* CR labeled tuples: test all of the above with top-level lets *)
+
+(* Labeled tuples nested in records *)
+
+let x = ref (~~(~x:1, ~y:2, ~x:3, 4))
+
+(* Good match *)
+let _1234 = match x with
+| { contents = ~~(~x=x0; ~y; ~x ; z)} -> (x0, y, x, z)
+[%%expect{|
+val x : (x:int * y:int * x:int * int) ref =
+  {contents = (~x:1, ~y:2, ~x:3, 4)}
+val _1234 : int * int * int * int = (1, 2, 3, 4)
+|}]
+
+(* Good partial match *)
+let _1  = match x with
+| { contents = ~~(~x; ..)} -> x
+[%%expect{|
+val _1 : int = 1
+|}]
+
+(* Empty match *)
+let () = match x with
+| { contents = ~~(..) } -> () 
+[%%expect{|
+|}]
+
+(* Wrong label *)
+let () = match x with
+| { contents = ~~(~w ; ..) } -> w 
+[%%expect{|
+Line 2, characters 15-26:
+2 | | { contents = ~~(~w ; ..) } -> w
+                   ^^^^^^^^^^^
+Error: This pattern matches values of type w:'a
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Missing unordered label *)
+let () = match x with
+| { contents = ~~(~x=x0; ~y ; ~x) } -> y 
+[%%expect{|
+Line 2, characters 15-33:
+2 | | { contents = ~~(~x=x0; ~y ; ~x) } -> y
+                   ^^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:'a * y:'b * x:'c
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Extra unordered label *)
+let () = match x with
+| { contents = ~~(~x=x0; ~y; ~x; w1; w2) } -> y 
+[%%expect{|
+Line 2, characters 15-40:
+2 | | { contents = ~~(~x=x0; ~y; ~x; w1; w2) } -> y
+                   ^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:'a * y:'b * x:'c * 'd * 'e
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Extra unordered label, open *)
+let () = match x with
+| { contents = ~~(~x=x0; ~y; ~x; w1; w2; ..) } -> y 
+[%%expect{|
+Line 2, characters 15-44:
+2 | | { contents = ~~(~x=x0; ~y; ~x; w1; w2; ..) } -> y
+                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:'a * y:'b * x:'c * 'd * 'e
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Missing label *)
+let () = match x with
+| { contents = ~~(~x=x0; ~y; x) } -> y 
+[%%expect{|
+Line 2, characters 15-31:
+2 | | { contents = ~~(~x=x0; ~y; x) } -> y
+                   ^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type x:'a * y:'b * 'c
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
+
+(* Extra label *)
+let () = match x with
+| { contents = ~~(~y=y0; ~y; ~x) } -> y 
+[%%expect{|
+Line 2, characters 15-32:
+2 | | { contents = ~~(~y=y0; ~y; ~x) } -> y
+                   ^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type y:'a * y:'b * x:'c
+       but a pattern was expected which matches values of type
+         x:int * y:int * x:int * int
+|}]
```

Comment:
Since this adds features that are dependent on whether things are principally known and which differ based on the presence of the principal flag, we should have some tests that show how that works out.  A couple that come to mind:
```ocaml
let f (z : ~~(x:_ * y:_)) =
  match z with
  | (~~(~y; ~x)) -> x + y
```
This works even when `-principal` is on because we have fully inferred the structure of the type, even if we don't know the types of the components.
```ocaml
let f = function (~~(~x; ~y)) -> x + y

let g z =
  (f z, match z with (~~(~y; ~x)) -> x + y)
```
This works without `-principal` but fails with it.  The reason is that when we check the second element of the (unlabeled) tuple, we can see the order of the (labeled) tuple, but that information hasn't been "fully" inferred yet.  It's a good illustration of what it means for something not to be principal - flipping the order of the pair makes this fail with both settings, and to be principal is, in part, to be robust to that kind of reordering.

(I think the expect test mechanism supports having different expected outputs based on the `-principal` flag.  For example, see `tests/typing-gadts/test.ml` and look for the word "Principal").

Reply:
Thanks for the extra tests! Added both, as well as the variant of the second where unlabeled tuple is flipped.